### ISSUE TITLE
Feature proposal: Signal client disconnect using `Request.signal`

### ISSUE THREAD

**Issue by brettwillis** [cloudflare/workerd#3033]
As I understand it, when a client disconnects, the context/"thread" for a request is aborted as is. And if `ctx.waitUntil()` was called, then those promises are allowed some time to resolve.

When you are running async callbacks using `ctx.waitUntil(promise)`, there is no easy way to know that the client has disconnected.

Conceptually, if you responded with a `ReadableStream`, then [controller.enqueue(chunk)](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamDefaultController/enqueue) should throw or [`writer.write(chunk)`](https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter/write) of the writable side should reject on the next write. This would require keeping the context alive with `ctx.waitUntil(writable.closed)` or something like that, otherwise the context would just die immediately anyway. However I've found that these don't seem to reliably throw/reject when the client disconnects. Perhaps that is a separate issue.

Much of what I'm describing above is from experience/testing or examining `workerd` rather than documentation, so feel free to correct me.

### Proposal

The [`Request.signal`](https://developer.mozilla.org/en-US/docs/Web/API/Request/signal) API is an existing standard API which reflects the `AbortSignal` passed when constructing the request. However the `signal` property of the incoming request currently does nothing.

I propose that the Workers runtime utilise this existing API by "aborting" the `signal` of the incoming request when a client disconnects. Any scripts that are interested in reacting when the client has disconnected can utilise this signal.

If an abort listener has been registered using `request.signal.addEventListener('abort', ...)` then the runtime will call those handler(s) when the client disconnects, before aborting the context. If the handler(s) synchronously call `ctx.waitUntil(promise)` then that will keep the context alive for a further period of time.

_Credit to uri2 [here](https://community.cloudflare.com/t/how-to-catch-handle-cancelled-requests/583797/3?u=arrowheadapps) for this inspiration of this idea, as other proposals I had entertained were new non-standard APIs such as `ctx.addEventListener('end', ...)` etc._

I'm happy to contribute if this proposal makes it through.

### Use cases

The use-cases I require this for are scenarios when the Worker is generating a long-running response stream on the fly, for example [Server-sent events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events).

When the client disconnects from the stream, it is useful to react to perform logging, or do cleanup such as signalling user presence in a database.

Currently the request will either die immediately, or if `ctx.waitUntil(writable.closed)` was called then it will continue running unknowingly until it gets terminated anyway 30s or so later.

**Comment by kentonv** [comment]
FWIW, I agree this is the right API.

(But I don't currently know when we'll be able to implement it.)

**Comment by brettwillis** [comment]
> (But I don't currently know when we'll be able to implement it.)

I'd be open to contributing this. But I wouldn't want to get started until the proposal has been through the appropriate reviews etc? Let me know.

**Comment by ricardopolo** [comment]
👍 to this feature, cloudflare should support abort requests

in the meantime, anyone have any workaround in mind?

**Comment by kentonv** [comment]
@mikenomitch @irvinebroque thoughts?

**Comment by irvinebroque** [comment]
We want to support `Request.signal`, and agree that registering the event listener on the incoming request makes sense. Matter of when not if in my mind (particularly now that the `cache` property of `Request` has some support in Workers https://github.com/cloudflare/cloudflare-docs/pull/17926)

Defer to others on if we think this is a viable thing for someone externally to contribute but in principle really appreciate that you're interested in contributing @brettwillis.

@jasnell FYI re: standards

**Comment by jasnell** [comment]
To be clear, we already support `request.signal` on the client side of a fetch, just not on the server side. We've had folks request the server-side for a while but it needs to come with the caveat that we likely cannot guarantee that it will fire in every case. Detecting that the connection has been aborted reliably can be a bit tricky but definitely doable.

One key question to determine is under what conditions would the `request.signal` be triggered on the server-side? Only when the runtime detects that the client has disconnected?

And just to make sure we're on the same page, this is what we're talking about, correct?

```js
export default {
  fetch(request) {
    request.signal.addEventListener('abort', () => {
      // If this is called, it means the incoming request was canceled and we should stop trying to do stuff for it.
    });
  }
}
```

**Comment by brettwillis** [comment]
> And just to make sure we're on the same page, this is what we're talking about, correct?

Yes correct. I'd add that the work in that callback may be async, and therefore involve `ctx.waitUntil()`:

```js
export default {
  fetch(request, env, ctx) {
    request.signal.addEventListener('abort', () => {
      // Do some logging or async work to signal other parts of the system
      const promise = ...;
      ctx.waitUntil(promise);
    });

    // Response may be a long-running readable stream
    const readable = ...;
    return readable;
  }
}
```

> One key question to determine is under what conditions would the request.signal be triggered on the server-side? Only when the runtime detects that the client has disconnected? Should it be called under other circumstances such as if the runtime cancels the request due to an error?

Good point, needs some careful consideration. Here's a train of thought from my side:

Is it correct to say that there are currently two ways for a request to end: (1) end "naturally" by emptying the event loop, or (2) "aborted" by the runtime by cancelling any pending callbacks in the event loop.

We could then say that `request.signal` is aborted whenever the runtime itself aborts a request as in (2) without it ending naturally? Worded in another way: "`abort` callbacks are guaranteed to be called when the runtime [knowingly] aborts a request before it naturally ends".

While this sounds clean, it may break down a little when reasoning about (3) and (4) below...

Conditions I can think of where the runtime currently aborts a request (cancelling event loop callbacks) are:
1. Platform detects a client disconnect (we see this as "Request cancelled" error in the logs). I think you are saying that the platform does not always reliably detect this, but when it does detect this the runtime does abort the request.
2. Internal reasons like server maintenance.
3. Timeout limits like callbacks in `ctx.waitUntil()` that take too long to resolve.
4. Resource limits like CPU or memory.

In the cases of (3) and (4), the script has already reached it's limits and therefore you might not want to give it more life with the `abort` callbacks?

There's also another case: I think "dangling" timers and async callbacks are cancelled from the event loop if they were not wrapped in `ctx.waitUntil()` once the response has been sent or an unhandled exception is thrown. In such cases the script has already yielded naturally under it's own control flow. It is fully capable of detecting it's own natural end or catching the exception, so I'd it does not need a signal from the runtime. We want the `abort` callback for when the runtime aborts the script outside of the script's control.

**Comment by jasnell** [comment]
> I think only case 1 makes sense to support via this API. AFAIK the platform does reliably detect client disconnects, so that's not an issue. It should be straightforward to implement.

I think this potentially becomes a bit wonky in deferred proxy cases. Specifically, if I have a case like...

```
export default {
  fetch(request) {
    request.signal.addEventListener('abort', () => {
      // should I expect this to be called?
    });
    return fetch('http://example/subrequest');
  }
}
```

In this case, because of deferred proxying, if the client disconnects after we've returned the response but before the response has been fully completed, would we expect the abort handler to be called? Do we care at this point?

**Comment by kentonv** [comment]
> In this case, because of deferred proxying, if the client disconnects after we've returned the response but before the response has been fully completed, would we expect the abort handler to be called? Do we care at this point?

I would say no, don't deliver any notification at that point. It would break the optimization, and I can't see why any app would need it.

**Comment by jasnell** [comment]
> I would say no, don't deliver any notification at that point. It would break the optimization, and I can't see why any app would need it.

We'll need to document that carefully then to avoid confusion. It would mean that the handler would be called in some cases after a response is sent but not called in others, depending on whether the deferred proxy optimization is in use or not... for instance, in a case like the following I would still expect the abort handler to be called since we're still running JavaScript after the response is returned. It just might be a bit confusing for folks if they aren't aware of when we apply the optimization or not.

```
export default {
  fetch(request) {
    request.signal.addEventListener('abort', () => {  /* ... */ });
    const rs = new ReadableStream({
      // ...
    });
    return new Response(rs);
  }
}
```

To be clear, I'm not disagreeing with you, just pointing out that we'll need to document this bit carefully.

**Comment by jasnell** [comment]
That's fair. And fwiw, this is what I meant earlier when I said, "we likely cannot guarantee that it will fire in every case." Specifically, there will be some cases (like deferred proxy) where we say that the abort listener will not be invoked even if the request has not been fully processed... simply because we might have passed the point at which we're actually running JavaScript. As long as that's understood this should be fairly straightforward to implement.

**Comment by ricardopolo** [comment]
Hey, any progress on this? We need to able to cancel requests in CF

**Comment by brettwillis** [comment]
> I think we might have to say: If the `signal` on an outgoing `fetch()` is the exact same `signal` from the original incoming request that started the Worker, it is ignored. If that feels wrong, then we would have to introduce a compat flag to fix it without breaking anyone. But I think it's fine really... if the client disconnects and there are no `waitUntil()`s, then all subrequests are canceled implicitly. So this only makes a difference if there are `waitUntil()`s, in which case you probably don't actually want to cancel the subrequest?

That's a really good point I hadn't thought of. It feels like most people would find it pretty unexpected for their subrequest to be cancelled because of the incoming requests `signal` when using `waitUntil()`s to follow up on the responses, and it would break a lot of workers (I know it would break some of mine), so would definitely have to be behind a flag. Even if I knew about this behaviour, it would probably keep catching me out in the future.

And personally it also feels a little convoluted that `signal` would be ignored if it's the same instance. I mean I can't actually think of a real circumstance when this would be unexpected, but it just smells a bit and makes me question now if `request.signal` is the right API... it seemed perfect until you mentioned this. Because aborting the incoming `request.signal` would be a new behaviour anyway, it wouldn't be breaking anything.

All that to say that of those two options my vote would be to implicitly ignore the same signal instance on subrequests. Either that or a non-standard API that lives beside `ctx.waitUntil()` like `ctx.addEventListener('end', ...)` or `ctx.whenCancelled(() => {...})` or similar. Are there any standards proposals?

> Maybe, but the same issue already exists today. Like if you use setInterval(), and then return a response eligible for deferred proxying, the interval will stop firing while the response streams through. Basically, once you've set up deferred proxying, JavaScript execution is canceled (unless you have waitUntil()).

Would we want `request.signal.addEventListener('abort', ...)` to behave the same as `ctx.waitUntil()`, i.e. in that adding the `abort` callback keeps JS alive can prevents the deferred proxy optimisation?

**Comment by kentonv** [comment]
A benefit of `ctx.addEventListener('end', ...)` is that it could work for non-fetch events, including RPC. But it would be non-standard, and being non-standard also means we will need to argue about exactly how the API should look. (Honestly the best thing about following standards... no bikeshedding.)

> Would we want request.signal.addEventListener('abort', ...) to behave the same as ctx.waitUntil(), i.e. in that adding the abort callback keeps JS alive can prevents the deferred proxy optimisation?

No I don't think so. I doubt most people would actually want this to prevent deferred proxying. We just won't deliver the abort event to JS if we've already gone into deferred proxying before the client disconnects. That's probably fine because this implies the JS context is already over and so all resources it was holding open have been closed implicitly (except for the proxy stream), so what else would it care to do?

**Comment by jasnell** [comment]
Why not `ctx.signal.addEventListener('abort')` ... that is, drop the `AbortSignal` off of the `ctx`? That would at least give us a standard API even if it is exposed via a non-standard property (`ctx`). 

> No I don't think so. I doubt most people would actually want this to prevent deferred proxying. We just won't deliver the abort event to JS if we've already gone into deferred proxying before the client disconnects. That's probably fine because this implies the JS context is already over and so all resources it was holding open have been closed implicitly (except for the proxy stream), so what else would it care to do?

+1 to this.

**Comment by kentonv** [comment]
> Why not ctx.signal.addEventListener('abort')

Good point, that also makes it easy to pass on the signal to other APIs expecting an AbortSignal. And it limits bikeshedding. :)

**Comment by brettwillis** [comment]
> Why not `ctx.signal.addEventListener('abort')`

Great idea, I like it too.

**Comment by jasnell** [comment]
Really comes down to prioritizing and scheduling the change here. It's on our roadmap. Difficult to say at the moment exactly when we'll be able to work on this.

**Comment by brettwillis** [comment]
Absolutely. Sorry I wasn't meaning "when", I was more asking, because I've offered to contribute to this, are there more reviews that need to take place, or is this a feature you'd prefer to keep in house?

**Comment by jasnell** [comment]
> ... or is this a feature you'd prefer to keep in house?

Certainly open to a contribution here. I just know there's likely some internal testing and possible an internal PR that would also be needed before we could pick this up in production so I think it at least warrants making sure we have someone on the runtime team engaged to work on this also.

**Comment by brettwillis** [comment]
Ok sounds great, then I'll wait to hear back before doing anything. Thanks 🙏🏼

**Comment by ricardopolo** [comment]
Hey! Any update o this?

**Comment by ricardopolo** [comment]
Hello Everyone.
Cloudflare team - any update in this important feature?

**Comment by jasnell** [comment]
No updates. It's in our list but it has not yet been scheduled for work.

**Comment by ricardopolo** [comment]
Any tip or workaround for managing a "Stop streaming" with this limitation? 

Using Cloudflare platform to interact LLMs (hosted in third party or with Cloudflare AI), when the LLM is streaming the response, this bug dont allow to cancel the generation of the message, and this is a very standard feature these days.

Any crazy idea as a workaround meanwhile? Thanks!

**Comment by huw** [comment]
I _think_ this doesn’t work if the response isn’t streaming (i.e. if the client disconnects while waiting for a response, rather than during a streamed response). For example:

```ts
export default {
   async fetch(request, env, ctx): Promise<Response> {
     request.signal.onabort = () => {
       console.log('The request was aborted!');
     };

     await scheduler.wait(10_000);
     return new Response(readable);
   },
 } satisfies ExportedHandler<Env>;
```

Should I be expecting this to work? This would be useful for me to perform cleanup if a client disconnects while waiting for a subrequest to return. (FWIW, the equivalent issue in Deno was denoland/deno#26761).

I can open a separate issue for this if you’d like :)

**Comment by npaun** [comment]
Hi @huw, I think the improvements to `Request.signal` will handle the case you described. It's similar to the test case I wrote here:

https://github.com/cloudflare/workerd/blob/92978adc483a94429a3f46b9d1ac4540ed35ab02/src/workerd/api/tests/request-client-disconnect.js#L127

However, unfortunately the new changes haven't yet been enabled in production. I'll post a reply in this issue to let you know when it's ready to use.

**Comment by kansson** [comment]
Hi @npaun, I'm also following this. Any timeline for when it will be released to production?

**Comment by npaun** [comment]
It's now live and working in production - I just checked. You'll need the compatibility flag `enable_request_signal` to use it for now.

**Comment by kansson** [comment]
> It's now live and working in production - I just checked. You'll need the compatibility flag `enable_request_signal` to use it for now.

The compatibility flag is not found during local development, but will it be available later?

**Comment by brettwillis** [comment]
Great work guys, thank you! Looked like it ended up being a bit of a mission in the end but good job.

**Comment by npaun** [comment]
@kansson Unfortunately it looks like wrangler is currently on an older version of workerd for unrelated reasons. 

If you need a workaround for local dev for now:

1. Grab a binary for your platform here: https://github.com/cloudflare/workerd/releases/latest
   - (Remember to gunzip it and mark it as executable)
 2. Run `export MINIFLARE_WORKERD_PATH=/path/to/workerd", etc. as described [here](https://github.com/cloudflare/workerd#local-worker-development-with-wrangler)
 3. wrangler will now use your copy of workerd instead of the bundled version.

**Comment by ricardopolo** [comment]
thanks cloudflare team, it looks working!
